From 948faf396cfd29d4d1b760d5210a6ffeb074bca4 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Wed, 6 Jan 2021 20:45:53 +0100
Subject: [PATCH 2/5] Set default compressor to zstd

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 mkinitcpio      | 2 +-
 mkinitcpio.conf | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/mkinitcpio b/mkinitcpio
index 9b585cd..38ed8be 100755
--- a/mkinitcpio
+++ b/mkinitcpio
@@ -508,7 +508,7 @@ if [[ $_optgenimg ]]; then
         die 'Unable to write to %s' "$_optgenimg"
     fi
 
-    _optcompress=${_optcompress:-${COMPRESSION:-gzip}}
+    _optcompress=${_optcompress:-${COMPRESSION:-zstd}}
     if ! type -P "$_optcompress" >/dev/null; then
         warning "Unable to locate compression method: %s" "$_optcompress"
         _optcompress=cat
diff --git a/mkinitcpio.conf b/mkinitcpio.conf
index ba1cce3..b07aa28 100644
--- a/mkinitcpio.conf
+++ b/mkinitcpio.conf
@@ -52,7 +52,7 @@ FILES=()
 HOOKS=(base udev autodetect modconf block filesystems keyboard fsck)
 
 # COMPRESSION
-# Use this to compress the initramfs image. By default, gzip compression
+# Use this to compress the initramfs image. By default, zstd compression
 # is used. Use 'cat' to create an uncompressed image.
 #COMPRESSION="gzip"
 #COMPRESSION="bzip2"
-- 
2.30.0.452.gfb7fa4a1fd

