From 0d162cc5136a14a1b921b203db7e066950c44d16 Mon Sep 17 00:00:00 2001
From: Chocobo1 <Chocobo1@users.noreply.github.com>
Date: Fri, 30 Oct 2020 12:20:28 +0800
Subject: [PATCH 4/5] Allow user to override default compression options

---
 mkinitcpio | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/mkinitcpio b/mkinitcpio
index 38ed8be..31d0d02 100755
--- a/mkinitcpio
+++ b/mkinitcpio
@@ -206,13 +206,13 @@ build_image() {
             msg "Creating %s-compressed initcpio image: %s" "$compress" "$out"
             ;;&
         xz)
-            COMPRESSION_OPTIONS+=('--check=crc32')
+            COMPRESSION_OPTIONS=('--check=crc32' "${COMPRESSION_OPTIONS[@]}")
             ;;
         lz4)
-            COMPRESSION_OPTIONS+=('-l')
+            COMPRESSION_OPTIONS=('-l' "${COMPRESSION_OPTIONS[@]}")
             ;;
         zstd)
-            COMPRESSION_OPTIONS+=('-19' '-T0')
+            COMPRESSION_OPTIONS=('-19' '-T0' "${COMPRESSION_OPTIONS[@]}")
             ;;
     esac
 
-- 
2.30.0.452.gfb7fa4a1fd

