From 5d7a86cd16abe89a080e41cb95a75a494ba0da83 Mon Sep 17 00:00:00 2001
From: Piotr Gorski <lucjan.lucjanov@gmail.com>
Date: Fri, 8 Jul 2022 18:31:51 +0200
Subject: [PATCH] Add new URL

Signed-off-by: Piotr Gorski <lucjan.lucjanov@gmail.com>
---
 uksmd.service | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/uksmd.service b/uksmd.service
index 9b81c7c..53a50af 100644
--- a/uksmd.service
+++ b/uksmd.service
@@ -1,6 +1,6 @@
 [Unit]
 Description=Userspace KSM helper daemon
-Documentation=https://gitlab.com/post-factum/uksmd
+Documentation=https://codeberg.org/pf-kernel/uksmd
 ConditionPathExists=/sys/kernel/pmadv/ksm
 
 [Service]
-- 
2.37.0.3.g30cc8d0f14

