From 3258ca1c632b00b174221b573e24048f0619223d Mon Sep 17 00:00:00 2001
From: Tobias Powalowski <tpowa@archlinux.org>
Date: Fri, 11 Mar 2022 15:47:59 +0100
Subject: [PATCH 12/24] add real symlink support

---
 functions | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/functions b/functions
index 36c773a..88bd563 100644
--- a/functions
+++ b/functions
@@ -574,7 +574,15 @@ add_file() {
     else
         quiet "adding file: %s" "$dest"
     fi
-    command install -Dm$mode "$src" "$BUILDROOT$dest"
+   #check if $src is a symlink
+    if [[ -L $src ]]; then
+        # add the symlink
+        add_file "$(realpath $src)"
+        # create the symlink
+        add_symlink "$src" "$(realpath $src)"
+    else
+       command install -Dm$mode "$src" "$BUILDROOT$dest"
+    fi
 }
 
 add_runscript() {
-- 
2.36.0.rc1

