From a375cda7e8eb39129829b0cecbd3f2f07c02e8ae Mon Sep 17 00:00:00 2001
From: Tobias Powalowski <tobias.powalowski@googlemail.com>
Date: Sun, 13 Mar 2022 20:46:26 +0100
Subject: [PATCH 19/20] Fix typo in commit reproducibility

---
 mkinitcpio | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mkinitcpio b/mkinitcpio
index ee01f0b..8a512ec 100755
--- a/mkinitcpio
+++ b/mkinitcpio
@@ -747,7 +747,7 @@ trap -- ERR
 
 # this is simply a nice-to-have -- it doesn't matter if it fails.
 ldconfig -r "$BUILDROOT" &>/dev/null
-# remove /var/cache/ldconfig/aux-cache for reproducability
+# remove /var/cache/ldconfig/aux-cache for reproducibility
 rm -f $BUILDROOT/var/cache/ldconfig/aux-cache
 
 # Set umask to create initramfs images and EFI images as 600
-- 
2.35.1.354.g715d08a9e5

