From 2ea8fe2261a15f680dcdaf4511445264aee11757 Mon Sep 17 00:00:00 2001
From: Simon Brand <simon.brand@postadigitale.de>
Date: Mon, 30 May 2022 12:35:14 +0000
Subject: [PATCH 27/27] autodetect does not work for overlay fs, set autodetect
 to failed

---
 install/autodetect | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/install/autodetect b/install/autodetect
index 846aa3a..6a576a6 100644
--- a/install/autodetect
+++ b/install/autodetect
@@ -31,7 +31,12 @@ build() {
 
     # detect filesystem for root
     if rootfstype=$(findmnt -uno fstype -T '/'); then
-        add_if_avail "$rootfstype"
+        if [ "${rootfstype}" = "overlay" ]; then
+            warning "cannot detect type of overlayfs root filesystem"
+            fs_autodetect_failed=1
+        else
+            add_if_avail "$rootfstype"
+        fi
     else
         error "failed to detect root filesystem"
         fs_autodetect_failed=1
-- 
2.37.0.rc0.15.g3b9a5a33c2

