From 457b1378a67374f4c5186f29346a35ffc8afb2c7 Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Fri, 21 Jul 2017 14:46:05 +0200
Subject: [PATCH] Consider also in_service_entity to state whether an entity is
 backlogged

---
 block/bfq-sched.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/block/bfq-sched.c b/block/bfq-sched.c
index da48db2..389a284d 100644
--- a/block/bfq-sched.c
+++ b/block/bfq-sched.c
@@ -1416,7 +1416,7 @@ static void bfq_deactivate_entity(struct bfq_entity *entity,
 			 */
 			bfq_update_next_in_service(sd, NULL);
 
-		if (sd->next_in_service) {
+		if (sd->next_in_service || sd->in_service_entity) {
 			/*
 			 * The parent entity is still backlogged,
 			 * because next_in_service is not NULL. So, no
@@ -1426,6 +1426,7 @@ static void bfq_deactivate_entity(struct bfq_entity *entity,
 			 * updated upwards.
 			 */
 			BUG_ON(sd->next_in_service == entity);
+			BUG_ON(sd->in_service_entity == entity);
 			break;
 		}
 
-- 
2.10.0

