From 06d2d8d95b2c7e5e1bdd3faa7e9533599ee6ca37 Mon Sep 17 00:00:00 2001
From: Paolo Valente <paolo.valente@linaro.org>
Date: Fri, 21 Jul 2017 12:08:57 +0200
Subject: [PATCH] block, bfq: reset in_service_entity if the pointed entity
 becomes idle

---
 block/bfq-sched.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/block/bfq-sched.c b/block/bfq-sched.c
index ad200f3..da48db2 100644
--- a/block/bfq-sched.c
+++ b/block/bfq-sched.c
@@ -1346,8 +1346,10 @@ static bool __bfq_deactivate_entity(struct bfq_entity *entity,
 
 	BUG_ON(is_in_service && entity->tree && entity->tree != &st->active);
 
-	if (is_in_service)
+	if (is_in_service) {
 		bfq_calc_finish(entity, entity->service);
+		sd->in_service_entity = NULL;
+	}
 
 	if (entity->tree == &st->active)
 		bfq_active_extract(st, entity);
-- 
2.10.0

